anndata/utils.py

```diff
@@ -59,6 +59,54 @@ def convert_to_dict_nonetype(obj: None):
     return dict()
 
 
+@singledispatch
+def dim_len(x, dim):
+    """\
+    Return the size of an array in dimension `dim`.
+
+    Returns None if `x` is an awkward array with variable length in the requested dimension.
+    """
+    return x.shape[dim]
+
+
+try:
+    from .compat import awkward as ak
+
+    @dim_len.register(ak.Array)
+    def dim_len_awkward(x, dim):
+        if dim == 0:
+            # dimension 0 is a special case - it is always of `ArrayType` and has a fixed length.
+            try:
+                return x.type.length
+            except AttributeError:
+                raise ValueError("The outermost type must be an `awkward.Array`!")
+        else:
+            arr_type = x.type
+            for _ in range(dim):
+                # we need to loop through the nested types for the other dimensions, e.g.
+                # ArrayType(RegularType(ListType(NumpyType('int64')), 200), 100)
+                try:
+                    arr_type = arr_type.content
+                except AttributeError:
+                    # RecordType and UnionType have multiple "contents" entries
+                    raise NotImplementedError(
+                        "This check is currently not implemented for RecordType and UnionType arrays. "
+                    )
+
+            try:
+                return arr_type.size
+            except AttributeError:
+                # the arrays is of variable length in the requested dimension
+                return None
```

Comment:
If we only support arrays aligned in one dimension we could simplify this to the `dim == 0` case and raise `NotImplementedError` otherwise. 

This would require some more custom code to test the `gen_awkward` function though: 
https://github.com/scverse/anndata/pull/647/files#diff-6fc7723d6e4e3fd25142614c3681bdda918de35d6c0a1ac18c3f57df7f4d7000R44-R63

What's your preferred solution @ivirshup?

Reply:
I think I would prefer for it to be simplified by following the pattern used in `awkward` here: https://github.com/scikit-hep/awkward/blob/main/src/awkward/_v2/operations/ak_to_regular.py

E.g. recurse down axes then and call `len` on the one specified.

I do think the tests could focus more on our actual use cases.

Reply:
Shall we go with the version proposed by Jim then + add some tests for it? 
https://github.com/scikit-hep/awkward/discussions/1654#discussioncomment-3521574

There's at least a chance that we could remove this in the future, in case awkward provides a similar function eventually.

Reply:
Yeah, I'd be fine with that.

I think it's pretty similar to `to_regular`, but uses `transform` instead of `recursively_apply`, right?

Reply:
In terms of how it works I'd say yes, but there's more special casing for the different array types than in `to_regular`, as far as i can tell.